ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.server;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// reference:
+// https://gist.github.com/danielflower/f54c2fe42d32356301c68860a4ab21ed
+// https://github.com/confluentinc/rest-utils/blob/master/core/src/main/java/io/confluent/rest/FileWatcher.java
+/**
+ * Watches a file and calls a callback when it is changed.
+ */
+public class FileWatcher implements Runnable {
+
+  private static final Logger log = LoggerFactory.getLogger(FileWatcher.class);
+
+  public interface Callback {
+    void run() throws Exception;
+  }
+
+  private volatile boolean shutdown;
+  private final WatchService watchService;
+  private final Path file;
+  private final Callback callback;
+  private Thread thread;
+
+  @SuppressFBWarnings(
+      value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE",
+      justification = "Null check on file.getParent() is present"
+  )
+  public FileWatcher(final Path file, final Callback callback) throws IOException {
+    this.file = Objects.requireNonNull(file);
+    Objects.requireNonNull(file.getParent(), "Watch location must have parent");
+    this.watchService = FileSystems.getDefault().newWatchService();
+    // Listen to both CREATE and MODIFY to reload, which handles delete then create.
+    file.getParent().register(watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_MODIFY);
+    this.callback = Objects.requireNonNull(callback);
+  }
+
+  /**
+   * Starts the file watcher in a separate thread
+   */
+  public synchronized void start() {
+    log.info("Starting file watcher to watch for changes: " + file);
+    thread = new Thread(this);
+    thread.start();
+  }
+
+  /**
+   * Stops watching the file and closes the watch service
+   */
+  public synchronized void shutdown() {
+    shutdown = true;
+    log.info("Stopped watching for TLS cert changes.");
+    if (thread != null) {
```

Comment:
I had the redundancy as extra insurance that we wouldn't leak resources (in case something changes with the FileWatcher code and shutdown() is no longer foolproof) but I think that was overkill. I've removed the interrupt() and simplified the Server code as you suggested.